kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+impl<'tcx> CoerceUnsizedIterator<'tcx> {
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        src_ty: Ty<'tcx>,
+        dst_ty: Ty<'tcx>,
+    ) -> CoerceUnsizedIterator<'tcx> {
+        CoerceUnsizedIterator { tcx, src_ty: Some(src_ty), dst_ty: Some(dst_ty) }
+    }
+}
+
+/// Iterate over the coercion path. At each iteration, it returns the name of the field that must
+/// be coerced, as well as the current source and the destination.
+/// E.g.: The first iteration of casting `Rc<String>` -> `Rc<&dyn Debug>` will return
+/// ```rust,ignore
+/// CoerceUnsizedInfo { field: "ptr", src_ty /* Rc<String> */, dst_ty /* Rc<&dyn Debug> */}
+/// ```
+/// while the last iteration will return:
+/// ```rust,ignore
+/// CoerceUnsizedInfo {
+///   field: None,
+///   src_ty: Ty, /* *const RcBox<String> */,
+///   dst_ty: Ty, /* *const RcBox<&dyn Debug> */
+/// }
+/// ```
+impl<'tcx> Iterator for CoerceUnsizedIterator<'tcx> {
+    type Item = CoerceUnsizedInfo<'tcx>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.src_ty.is_none() {
+            assert_eq!(self.dst_ty, None, "Expected no dst type.");
+            return None;
+        }
+
+        // Extract the pointee types from pointers (including smart pointers) that form the base of
+        // the conversion.
+        let src_ty = self.src_ty.take().unwrap();
+        let dst_ty = self.dst_ty.take().unwrap();
+        let field = match (&src_ty.kind(), &dst_ty.kind()) {
+            (&ty::Adt(src_def, src_substs), &ty::Adt(dst_def, dst_substs)) => {
+                // Handle smart pointers by using CustomCoerceUnsized to find the field being
+                // coerced.
+                assert_eq!(src_def, dst_def);
+                let src_fields = &src_def.non_enum_variant().fields;
+                let dst_fields = &dst_def.non_enum_variant().fields;
+                assert_eq!(src_fields.len(), dst_fields.len());
+
+                let CustomCoerceUnsized::Struct(coerce_index) =
+                    custom_coerce_unsize_info(self.tcx, src_ty, dst_ty);
+                assert!(coerce_index < src_fields.len());
+
+                self.src_ty = Some(src_fields[coerce_index].ty(self.tcx, src_substs));
+                self.dst_ty = Some(dst_fields[coerce_index].ty(self.tcx, dst_substs));
+                Some(src_fields[coerce_index].name)
+            }
+            _ => {
+                // Base case is always a pointer (Box, raw_pointer or reference).
+                assert!(
+                    extract_pointee(src_ty).is_some(),
+                    "Expected a pointer, but found {src_ty:?}"
+                );
+                None
+            }
+        };
+        Some(CoerceUnsizedInfo { field, src_ty, dst_ty })
+    }
+}
+
+/// Get information about an unsized coercion.
+/// This code was extracted from `rustc_monomorphize` crate.
+/// <https://github.com/rust-lang/rust/blob/4891d57f7aab37b5d6a84f2901c0bb8903111d53/compiler/rustc_monomorphize/src/lib.rs#L25-L46>
```

Comment:
Should we have a tracking issue to boost this into main rustc?

Reply:
We could, but it's fairly low in our priority list.